### PR TITLE
Multiprocess

### PR DESCRIPTION
**Pull request by JingJZ160**
Description
Turning on multiprocess in the ParallelProcessor. To enable this, we have to:

replace all local/inner functions/lambdas/classes with public functions/classes.
replace metric decorator with modified MetricWrapper
Note that turning on parallel processing on small dataset is actually counterproductive since the overhead of spawning processes is high. In this case it's better to use single process. For now we add an option to let user select whether to use multiprocess. In the future we may want to let the framework decide what to do.

### CODE REVIEW

File: foreshadow/parallelprocessor.py [hunk 1]
```diff
@@ -403,12 +468,16 @@ def fit_transform(self, X, y=None, **fit_params):
...
         )
-
         if not result:
             # All transformers are None
             return X[[]]
 
         Xs, transformers = zip(*result)
+        if update_column_sharer:
+            self._update_original_column_sharer(column_sharer, transformers)
+            self._update_transformers_with_updated_column_sharer(
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/parallelprocessor.py]: You are probably right. I was in a train of thought that we need to update CS everywhere. Let me verify that the CS instance is auto-updated in the transformers.

File: foreshadow/tests/test_transformers/test_transformers.py [hunk 2]
```diff
@@ -130,6 +130,120 @@ def test_transformer_parallel():
...
+    assert Xs.equals(df)
+    assert len(cs["intent"]) == len(list(df.columns.values))
+    assert (
+        cs["intent", "crim"] == "Numeric"
+        and cs["intent", "zn"] == "Categoric"
+        and cs["intent", "indus"] == "Categoric"
+    )
+
+
+def test_transformer_multiprocess_smart_transformers_update_column_sharer():
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/tests/test_transformers/test_transformers.py]: hmm actually I pick this name on purpose. Do you have any suggestions?

File: foreshadow/tests/test_transformers/test_transformers.py [hunk 3]
```diff
@@ -130,6 +130,120 @@ def test_transformer_parallel():
...
+            ("group3", StandardScaler(), ["indus"]),
+        ],
+        n_jobs=-1,
+        collapse_index=True,
+    )
+
+    proc.fit_transform(df)
+    assert len(cs) == 0
+
+
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/tests/test_transformers/test_transformers.py]: Save as above.

File: foreshadow/tests/test_transformers/test_transformers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/tests/test_transformers/test_transformers.py]: Ah I got it reversed...will update.

File: foreshadow/parallelprocessor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/parallelprocessor.py]: The reason that I have this line is because the `transformers` extracted from the `result` are returned from the parallel process, which means that are deep copies of the original transformers (and in turn copies of the original column_sharer). The cs instances have different object ids comparing to the original one. Given that we are updating the `self.transformer_list` using the `transformers` in the next line with `self._update_transformer_list(transformers)`, we will need to update the cs instances in the `transformers` first.
